contracts/strategies/allocation/AllocationWithVotingBase.sol

```diff
@@ -0,0 +1,306 @@
+    error UNAUTHORIZED();
+    error INVALID_TIME();
+
+    event ApplicationSubmitted(
+        uint256 indexed applicationId, bytes32 indexed identityId, address recipient, Metadata metadata, address _sender
+    );
+
+    event ApplicationStatusUpdated(address indexed applicant, uint256 indexed applicationId, Status status);
+    event ApplicationVoted(
+        address indexed voter, uint256 indexed applicationId, uint256 vote, address token, address recipient
+    );
+    event TimestampsUpdated(
+        uint256 applicationStartTime, uint256 applicationEndTime, uint256 votingStartTime, uint256 votingEndTime
+    );
+
+    /// @notice Struct to hold details of an application
+    struct Application {
+        bytes32 identityId;
+        address recipient;
+        uint256 applicationId;
+        uint256 amount;
+        uint256 votes;
+        Status status;
+        Metadata metadata;
+        address creator;
+    }
+
+    // Note: this is mapped to the Allo global status's in the mapping below.
+    /// @notice Enum for the local status of the application
+    enum Status {
+        PENDING,
+        ACCEPTED,
+        REJECTED,
+        REAPPLIED
+    }
+
+    /// @notice applicationId - Status
+    mapping(uint256 => ApplicationStatus) public applicationStatus;
+
+    /// @notice applicationId -> Application
+    mapping(uint256 => Application) public applications;
+
+    modifier onlyAllo() {
+        if (msg.sender != address(allo)) {
+            revert NOT_ALLO();
+        }
+        _;
+    }
+
+    /// @notice Initialize the contract
+    /// @param _data The data to be decoded
+    function initialize(bytes memory _data) public initializer {
+        (address _allo, address _registry, bool _identityRequired) = abi.decode(_data, (address, address, bool));
+        allo = Allo(_allo);
+        registry = Registry(_registry);
+        identityRequired = _identityRequired;
+
+        // TODO: Ask who calls this?  is it allo.sol or deployer ?
+    }
+
+    /// @notice Returns the owner identity
+    function getOwnerIdentity() external view override returns (string memory) {
+        // TODO: ASK zach
+    }
+
+    /// @notice Applies to the pool
+    /// @param _data The data to be decoded
+    /// @param _sender The sender of the transaction
+    function applyToPool(bytes memory _data, address _sender) external payable override onlyAllo returns (uint256) {
+        if (applicationStartTime <= block.timestamp || applicationEndTime >= block.timestamp) {
+            revert APPLICATIONS_NOT_OPEN();
+        }
+
+        (uint256 applicationId, address recipient, bytes32 identityId, Metadata memory metadata) =
+            abi.decode(_data, (uint256, address, bytes32, Metadata));
+
+        if (identityRequired && !registry.isOwnerOrMemberOfIdentity(identityId, _sender)) {
+            revert UNAUTHORIZED();
+        }
+
+        // Note: applicationId is 0 when the application is new
+        if (applicationId == 0) {
```

Comment:
Should we add a check, if there is already a application for the identityId?
With check: only one application by one identity
Without check: many applications by one identity

 [spec](https://github.com/allo-protocol/allo-v2/pull/31/files#r1247589613)

@nfrgosselin

Reply:
We should add a check !